flake.nix

```
@@ -158,16 +161,26 @@
               "runtime-tests"
               "composablejs"
             ];
-            file-blacklist = [ "flake.nix" "flake.lock" ];
+            file-blacklist = [
+              # does not makes sence to black list,
+              # if we changed some version of tooling(seldom), we want to rebuild
+              # so if we changed version of tooling, nix itself will detect invalidation and rebuild
+              # "flake.lock"
+            ];
           in lib.cleanSourceWith {
             filter = lib.cleanSourceFilter;
             src = lib.cleanSourceWith {
               filter = let
                 customFilter = name: type:
-                  (!(type == "directory"
-                    && builtins.elem (baseNameOf name) dir-blacklist))
-                  && (!(type == "regular"
-                    && builtins.elem (baseNameOf name) file-blacklist));
+                  !((type == "directory"
+                    && builtins.elem (baseNameOf name) directory-blacklist)
+                    || (type == "regular"
+                      && builtins.elem (baseNameOf name) file-blacklist)
+                    # assumption that nix is final builder, 
+                    # so there would no be sandwitch like  .*.nix <- build.rs <- *.nix
+                    # and if *.nix changed, nix itself will detect only relevant cache invalidations 
+                    || (type == "regular"
+                      && lib.strings.hasSuffix ".nix" name));
```

```suggestion
                  !(
                    (type == "directory" && builtins.elem (baseNameOf name) directory-blacklist)
                    || (type == "regular" && builtins.elem (baseNameOf name) file-blacklist)
                    # assumption that nix is final builder, 
                    # so there would no be sandwitch like  .*.nix <- build.rs <- *.nix
                    # and if *.nix changed, nix itself will detect only relevant cache invalidations 
                    || (type == "regular" && lib.strings.hasSuffix ".nix" name)
                  );
```

---

imho, so is simpler to reading

---

or you can use [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) and rewrite it so
```suggestion          
                    !(type == "directory" && builtins.elem (baseNameOf name) directory-blacklist)
                    && !(type == "regular" && builtins.elem (baseNameOf name) file-blacklist)
                    # assumption that nix is final builder, 
                    # so there would no be sandwitch like  .*.nix <- build.rs <- *.nix
                    # and if *.nix changed, nix itself will detect only relevant cache invalidations 
                    && !(type == "regular" && lib.strings.hasSuffix ".nix" name));
```

---

image you have
```
negative-filter
positive-filter
```

```
positive-filter
negative-filter
```

negative filter matches positively as negative. so both match for true, but then result either rejected or accepted. 

i find that for natural - kind of LINQ, |> pipes, | pipes, chain.of.iter.map.split.join so.

---

you can pick negative filter and combine with other negative filter which both are positive for eventual negativity. true || true.  and then parent decides how to apply filter. while you say that filter itself decides kind of.

---

you tell imperstively

if A not x and not y and not z

i tell
negative = x | y | z
that is type

then  make new type

filter = reject negative abort | accept positive 
 where positive is identity for now

it seems more type and composable.  

imageni docker, git, nix, codeowners, got composition of filters with proper folder hierarchies. you can grow mono repo huge like that.

---

i guess you will find such a lib in haskell or can write in nix))

---

It is the same for me
ok, let's use !(bool || bool || bool)